### PR TITLE
Handle transient state errors in `RedshiftResumeClusterOperator` and `RedshiftPauseClusterOperator`

### PR DESCRIPTION
**Pull request by syedahsn**
There were some issues with transient state changes with the `RedshiftResumeClusterOperator` and the `RedshiftPauseClusterOperator`, which resulted in errors being thrown, and causing system tests to fail. This PR solves the issue by allowing multiple attempts to pause or resume the cluster if desired.

Edit:
This PR also contains a minor fix on the `get_cluster_snapshot_status` function, which is to remove a redundant parameter `cluster_identifier`, which was originally being passed to the `describe_cluster_snapshots` function. The function only accepts one of the two parameters currently provided: either the `snapshot_identifier` or the `cluster_identifier`. Passing both to the function raises an `InvalidParameterCombination` exception.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ def create_cluster_snapshot(
         )
         return response["Snapshot"] if response["Snapshot"] else None
 
-    def get_cluster_snapshot_status(self, snapshot_identifier: str, cluster_identifier: str):
+    def get_cluster_snapshot_status(self, snapshot_identifier: str):
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/hooks/redshift_cluster.py]: I feel changing definition of a hook method is a breaking change

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py [hunk 2]
```diff
@@ -166,7 +166,6 @@ def get_cluster_snapshot_status(self, snapshot_identifier: str, cluster_identifi
         """
         try:
             response = self.get_conn().describe_cluster_snapshots(
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/hooks/redshift_cluster.py:178]: why this change, I think the purpose of this hook method is to get the snapshot of a cluster WDYT?

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py (hunk 1 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/hooks/redshift_cluster.py]: Good point. The reason I removed the `cluster_identifier` is because the `describe_cluster_snapshots` function throws an `InvalidParameterCombination` when both the `cluster_identifier` and `snapshot_identifier` are provided. 
```
botocore.exceptions.ClientError: An error occurred (InvalidParameterCombination) when calling the DescribeClusterSnapshots operation: Cannot specify both a Cluster identifier and a snapshot identifier.
```
I'll add the `cluster_identifier` back to the function definition, and add a deprecation warning to it.

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py (hunk 2 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/hooks/redshift_cluster.py:178]: See above comment. Basically, the `describe_cluster_snapshots` function requires only one of the 2 parameters.

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py [hunk 3]
```diff
@@ -157,16 +158,24 @@ def create_cluster_snapshot(
...
 
-    def get_cluster_snapshot_status(self, snapshot_identifier: str, cluster_identifier: str):
+    def get_cluster_snapshot_status(self, snapshot_identifier: str, cluster_identifier: str | None = None):
         """
         Return Redshift cluster snapshot status. If cluster snapshot not found return ``None``
 
         :param snapshot_identifier: A unique identifier for the snapshot that you are requesting
-        :param cluster_identifier: The unique identifier of the cluster the snapshot was created from
+        :param cluster_identifier: (deprecated) The unique identifier of the cluster
+            the snapshot was created from
```
Comment (eladkal) [inline airflow/providers/amazon/aws/hooks/redshift_cluster.py:167]: Can you please share some context?
The PR description says the motivation is issues with system tests but there should be more to it?
Why are we deprecating this parameter? If boto3 accept this as valid input why should we prevent users from using it?

File: airflow/providers/amazon/aws/operators/redshift_cluster.py [hunk 4]
```diff
@@ -392,20 +393,31 @@ def __init__(
...
+                redshift_hook.get_conn().resume_cluster(ClusterIdentifier=self.cluster_identifier)
+                return
+            except redshift_hook.get_conn().exceptions.InvalidClusterStateFault as error:
+                self.attempts = self.attempts - 1
+
+                if self.attempts > 0:
+                    self.log.error("Unable to resume cluster. %d attempts remaining.", self.attempts)
+                    time.sleep(self.attempt_interval)
+                else:
+                    raise error
```
Comment (eladkal) [inline airflow/providers/amazon/aws/operators/redshift_cluster.py]: This is essentially the behavior of `retries` and `retry_delay` of BaseOperator

Why do we need another layer of "internal retries"?
I can argue that if we have such need then it possibly goes beyond this single operator and may apply to others.
If we must have it lets not expose the parameters to the users. I'm not sure if we want each indvidual operator to set it's own user facing logic for retry.

I think this is something should be discussed in more depth. I'd like to hear others on this one.

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py (hunk 3 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/hooks/redshift_cluster.py:167]: The change was made as a result of debugging during testing for the redshift system test. It was not originally in scope of the refactoring of the `RedshiftResumeClusterOperator` and `RedshiftPauseClusterOperator`. I included it in this PR because the `get_cluster_snapshot_status` function does not work as written.

>If boto3 accept this as valid input why should we prevent users from using it?

Boto3 does not accept both inputs at the same time. If both are provided, we get an `InvalidParameterCombination` exception.

File: airflow/providers/amazon/aws/operators/redshift_cluster.py (hunk 4 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/redshift_cluster.py]: The reason the retries are necessary is because it's not clear why the cluster initially fails to resume (or pause). Despite the state of the cluster being in the correct state to perform the operation, the first few attempts fail (depending on the `attempt_interval`). 
You have a good point about the behaviour of `retries` and `retry_delay`. Should I just use those, and remove `attempts` and `attempt_interval`?
> If we must have it lets not expose the parameters to the users. I'm not sure if we want each indvidual operator to set it's own user facing logic for retry.

I'm not sure what you mean by this.

Comment (o-nikolas) [reply airflow/providers/amazon/aws/operators/redshift_cluster.py]: IMHO the top level `retries` and `retry_delay` are for users to generally dictate what an operator should do when it unexpectedly fails. The code Syed is adding here is to actually account for a known poor behaviour on the AWS redshift API/SDK where it unreliably reports a good state but still rejects mutations to the cluster for a short time.

Though I do agree with @eladkal's second point, we don't need to expose the parameters to the users. @syedahsn we can drop the arguments you added to the method stub and just define them locally here with reasonable defaults.

Comment (eladkal) [reply airflow/providers/amazon/aws/operators/redshift_cluster.py]: > The code Syed is adding here is to actually account for a known poor behaviour on the AWS redshift API/SDK where it unreliably reports a good state but still rejects mutations to the cluster for a short time.

I agree. I see this in same way as we handle random http failures for example:
https://github.com/apache/airflow/blob/3d5f34cb0f294d21dd1ba244af0fa06873377f11/airflow/providers/google/cloud/hooks/dataprep.py#L77-L78

File: airflow/providers/amazon/aws/hooks/redshift_cluster.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/redshift_cluster.py:167]: In that case, maybe leave the parameter in and add a block at the beginning of the method along the lines of 

```
if not(cluster_identifier ^ snapshot_identifier):
    raise SomeException("cluster_identifier or snapshot_identifier must be included, but not both")
```    

The ^ is an XOR operator, which returns true if one and only one of those exists.

```
In [1]: t = True ; f = False

In [2]: t ^ f
Out[2]: True

In [3]: f ^ t
Out[3]: True

In [4]: t ^ t
Out[4]: False

In [5]: f ^ f
Out[5]: False
```

File: airflow/providers/amazon/aws/operators/redshift_cluster.py (hunk 4 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/operators/redshift_cluster.py]: > > The code Syed is adding here is to actually account for a known poor behaviour on the AWS redshift API/SDK where it unreliably reports a good state but still rejects mutations to the cluster for a short time.
> 
> I agree. I see this in same way as we handle random http failures for example:
> 
> https://github.com/apache/airflow/blob/3d5f34cb0f294d21dd1ba244af0fa06873377f11/airflow/providers/google/cloud/hooks/dataprep.py#L77-L78

I really like the `@retry` decorator approach too, we should use that more often within airflow codebase.

File: airflow/providers/amazon/aws/operators/redshift_cluster.py [hunk 5]
```diff
@@ -397,15 +396,27 @@ def __init__(
         super().__init__(**kwargs)
         self.cluster_identifier = cluster_identifier
         self.aws_conn_id = aws_conn_id
+        # These parameters are added to address an issue with the boto3 API where the API
+        # prematurely reports the cluster as available to receive requests. This causes the cluster
+        # to reject initial attempts to resume the cluster despite reporting the correct state.
+        self.attempts = 10
+        self.attempt_interval = 15
```
Comment (eladkal) [inline airflow/providers/amazon/aws/operators/redshift_cluster.py]: should we mark as private?
since these are not user facing I'd rather users not to relay on this (In case they subclass the operator)

Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/redshift_cluster.py]: good point

**Comment by potiuk** [comment]
LGTM. @eladkal ?

**Review by eladkal — APPROVED** [review]
